File: c#/_57_SquareRootConvergents.cs
using System;
using System.Numerics;

public class _57_SquareRootConvergents{
	public static int Euler57(){
		int res = 0;
		BigInteger num = 1, den = 1;
		for (int i = 1; i <= 1000; i++){
			BigInteger newNum = num + 2 * den;
			BigInteger newDen = num + den;
			num = newNum;
			den = newDen;
			
			if (CountDigits(num) > CountDigits(den)){
				res++;
			}
		}
		return res;
	}
	
	public static int CountDigits(BigInteger num){
		int count = 0;
		while (num > 0){
			num /= 10;
			count++;
		}
		return count;
	}
	
	public static void Main(){
		Console.WriteLine(Euler57());
	}
}

File: java/_57_SquareRootConvergents.java
import java.util.*;
import java.math.BigInteger;

public class _57_SquareRootConvergents{
    
    public static void main(String[] args){
        // Iteration 1: 1 + 1/2 = 3/2
        // Iteration 2: 1 + 1/(2 + 1/2) = 1 + 1/(5/2) = 1 + 2/5 = 7/5
        // Iteration 3: 1 + 1/(2 + 1/(2+1/2)) = 1 + 1/(12/5) = 1 + 5/12 = 17/12
        
        // Let the result of iteration x be n/d.
        // Then iteration (x+1) yields 1 + 1/(1 + n/d) = (2 + n/d)/(1 + n/d).
        // Multiplying both parts of the fraction by d results in (2*d + n)/(d + n).
        // Thus, n becomes (2*d + n) and d becomes (d + n) with each subsequent iteration.
        
        int numFractions = 0;
        BigInteger num = BigInteger.valueOf(3);
        BigInteger den = BigInteger.valueOf(2);
        
        for (int i = 0; i < 999; i++){
            BigInteger newNum = num.add(den.multiply(BigInteger.valueOf(2)));
            BigInteger newDen = num.add(den);
            
            // System.out.println(newNum + "/" + newDen);
            if (newNum.toString().length() > newDen.toString().length()){
                numFractions++;
            }
            num = newNum;
            den = newDen;
        }
        
        System.out.println("# of fractions with their numerator having more digits than the denominator: " + numFractions); // 153
    }
}